### PR TITLE
Catch-up after Federation Outage (split, 1)

### PR DESCRIPTION
**Pull request by reivilibre**
This adds and keeps updated `destination_rooms`, a table used for determining the latest messages for each (destination, room) pair.

### CODE REVIEW

File: synapse/storage/databases/main/schema/delta/58/14_catchup_destination_rooms.sql
Comment (richvdh) [inline synapse/storage/databases/main/schema/delta/58/14_catchup_destination_rooms.sql:1]: we've already got a 14: please can you pick a new name?

File: synapse/storage/databases/main/schema/delta/58/14_catchup_destination_rooms.sql
Comment (richvdh) [inline synapse/storage/databases/main/schema/delta/58/14_catchup_destination_rooms.sql:18]:
```suggestion
-- This stores, for each (destination, room) pair the stream_ordering of the
```

File: synapse/storage/databases/main/transactions.py
Comment (richvdh) [inline synapse/storage/databases/main/transactions.py:281]: event_id seems unused?

File: synapse/storage/databases/main/transactions.py
Comment (richvdh) [inline synapse/storage/databases/main/transactions.py:317]: comment please. why are we doing this?

**Review by richvdh — CHANGES_REQUESTED** [review]
this is going to need a newsfile.

File: synapse/storage/databases/main/transactions.py
Comment (reivilibre) [reply synapse/storage/databases/main/transactions.py]: Is this OK?

File: synapse/storage/databases/main/purge_events.py
Comment (richvdh) [inline synapse/storage/databases/main/purge_events.py:205]: well, destination_rooms doesn't contain events. can you reword this please?

File: synapse/storage/databases/main/purge_events.py
Comment (richvdh) [inline synapse/storage/databases/main/purge_events.py:210]: destination_rooms doesn't have an index on `stream_ordering`, so this is going to be very inefficient.

Can we give `destination_rooms` an index on room_id?

File: synapse/storage/databases/main/transactions.py
Comment (richvdh) [inline synapse/storage/databases/main/transactions.py:168]: I'm slightly struggling to see how this is related. Can you help me out?

File: synapse/storage/databases/main/transactions.py
Comment (richvdh) [inline synapse/storage/databases/main/transactions.py:287]: there is no event_id?

**Review by richvdh — CHANGES_REQUESTED** [review]

File: synapse/storage/databases/main/transactions.py
Comment (reivilibre) [reply synapse/storage/databases/main/transactions.py:168]: `retry_last_ts` may be NULL — this was already the case before as far as the schema goes, but otherwise this is untrodden until now.

(This case is hit when we insert a `destinations` row to satisfy the constraint that `destination_rooms` has.)

File: synapse/storage/databases/main/purge_events.py
Comment (reivilibre) [reply synapse/storage/databases/main/purge_events.py:210]: Following you about the missing index, but are you saying you want both indices on `stream_ordering` and `room_id`, or just the latter (and then to reword the query to use it)?

Comment (richvdh) [reply synapse/storage/databases/main/purge_events.py:210]: just the latter. Hopefully it will need less index updating than an index on stream_ordering.

File: synapse/storage/databases/main/transactions.py
Comment (richvdh) [reply synapse/storage/databases/main/transactions.py:168]: got it.

`result["retry_last_ts"]` can never be negative, so I'd omit the last clause.

File: synapse/storage/databases/main/purge_events.py
Comment (richvdh) [reply synapse/storage/databases/main/purge_events.py:210]: btw: I think the `purge_room` api might also need an update (https://github.com/matrix-org/synapse/blob/master/docs/admin_api/purge_room.md)

Comment (reivilibre) [reply synapse/storage/databases/main/purge_events.py:210]: Is this better?

File: synapse/storage/databases/main/purge_events.py
Comment (richvdh) [inline synapse/storage/databases/main/purge_events.py:212]: now that stream_ordering isn't an FK, I think this is redundant.

File: synapse/storage/databases/main/schema/delta/58/15_catchup_destination_rooms.sql
Comment (richvdh) [inline synapse/storage/databases/main/schema/delta/58/15_catchup_destination_rooms.sql:30]: ... also: it's redundant. The point is that stream_ordering doesn't actually *need* to refer to a specific event; if the event in question gets deleted, it doesn't impact the validity of the stream_ordering here.

File: synapse/storage/databases/main/schema/delta/58/15_catchup_destination_rooms.sql
Comment (richvdh) [inline synapse/storage/databases/main/schema/delta/58/15_catchup_destination_rooms.sql:34]: might be worth adding a comment to say why this is needed.

**Review by richvdh — CHANGES_REQUESTED** [review]

**Review by richvdh — APPROVED** [review]
lgtm, but as per #synapse-dev: could you hold off merging it for now?

File: changelog.d/8230.misc
Comment (richvdh) [inline changelog.d/8230.misc:1]: once you land the other parts of this feature, it'll be good to update this so that the changelog entries get combined.

**Pull request merged**
